docs/proposals/25-private-cloud-profiles.md

```diff
@@ -0,0 +1,299 @@
+
+### Current State
+
+Cloud Profiles are non-namespaced resources. This means that in a typical gardener installation, only a handful of cloud profiles (i.e. one per cloud provider) exist. They can be consumed by any project's clusters. Consequently, when a project requires changes to cloud profiles for any of the reasons mentioned in the [context section](#context), they are changed for the entire gardener landscape. Since some projects might require frequent changes to a cloud profile, this become quite a cumbersome process. Additionally, it adds confusion to every project's workflow if say for example a team requires an extension of the expiration date of a Kubernetes version since they will be wondering why it was extended.
+
+### Goals
+
+- Reduce load on Gardener operators for maintaining cloud profiles
+- Possibly reduce the time a team has to wait until a change in a cloud profile is reflected in production
+- Make it so that project-scoped information in the cloud profile is only visible to the relevant project
+- Full backward compatibility
+
+### Non-Goals
+
+- Automate approval process for changes in Gardeners cloud profiles
+- Add different, possibly unsupported Kubernetes versions or machine image versions and names
+
+## Proposal
+
+It is proposed to implement a solution that enables project administrators to create custom (private) cloud profiles. These cloud profiles would be consumed by project users of the project they were created in.
+
+### Approach
+
+First of all, a new, namespaced API object is defined, the `PrivateCloudProfile`. Its type definition is very similar to the `CloudProfile` object but it omits the `type`, `providerConfig` and `seedSelector` fields. The `type` field is omitted as it is inherited from the parent cloud profile. The `providerConfig` field is omitted as gardener itself does not know anything about the structure of this field and cannot merge it accordingly. The `seedSelector` field is omitted as it is too complex for this GEP but may be added in the future.
+
+The general approach is that a private cloud profile inherits from a cloud profile using a `parent` field in the private cloud profile and should therefore not change the type or provider config. Fields such as `machineTypes`, `volumeTypes`, `regions` and `caBundle` are going to be merged with the parent cloud profile. However, a restriction need to be defined so that the `kubernetes` and `machineImages` fields in a private cloud profile may only be adjusted by a gardener operator to reduce the chance of a team staying on an unsupported kubernetes version. A similar problem is already solved in Gardener using custom RBAC verbs [here](https://github.com/gardener/gardener/blob/master/plugin/pkg/global/customverbauthorizer), the approach is described in more detail in the [custom RBAC verb section](#custom-rbac-verb).
```

Comment:
I have few concerns about supporting old Kubernetes version- 
- How dropping support for old Kubernetes version will look like, once the kubernetes version support is dropped, shoot of that Kubernetes version will not work.
- How long/extra do we have to wait to drop support of old Kubernetes version, user can ask to extend the life of Kubernetes version for 6 months/1 year, will gardener have to wait for it?
- In case of old Kubernetes support is removed from gardener, and the private cloud profile still has old Kubernetes, will we upgrade all shoot forcefully and remove old Kubernetes version from private cloud profile?

If above is the case, how it will be different from the current approach?

Reply:
I have been made aware that some teams are not immediately ready to upgrade to the next version of Kubernetes when the expiration date is due. Therefore, I think, we currently have to extend the expiration date in the cloud profile leaving every project open to using the "outdated, but extended" version. With private cloud profiles we could extend the expiration date on a project level and then figure with the projects team how long the preparation for the next version will take them. However, enabling this extended expiration date of a Kubernetes version should be the exception not the norm. Once all the extended Kubernetes versions have expired, Gardener can drop support for that version. It differs from our current approach in that it can be done on a project basis and not for all projects. Is that what you were asking about?

Reply:
In general, handling Kubernetes version updates is not part of this GEP and not related to private cloud profiles. The only thing that would change with private cloud profiles is that the Kubernetes version could be extended on a per project basis. But a new upgrade strategy will not be introduced as of this GEP.

Reply:
>With private cloud profiles we could extend the expiration date on a project level and then figure with the projects team how long the preparation for the next version will take them. However, enabling this extended expiration date of a Kubernetes version should be the exception not the norm. Once all the extended Kubernetes versions have expired, Gardener can drop support for that version.

It seems more work to me than what is being done currently. 
Current scenario - 
- Operator communicates with teams to extend the lifetime of the Kubernetes version.
- Operator can manage one place stuff like how long a certain Kubernetes version's life should be increased and expire the version accordingly.

New scenario(this GEP) - 
- Operator communicates with teams to extend the lifetime of the Kubernetes version.
- Accordingly, add the Kubernetes version in the respective private cloudprofiles and remove it from the main CloudProfile.
- This may be required to be done for multiple projects.
- Then manage all private cloudprofiles to remove the old Kubernetes version. (Again, it can be multiple cloudprofiles).

Overall more work is required than what is being done today with little to no benefit.

Reply:
With the new case, some teams who would not have needed PrivateCludprofile will create it only for using the old Kubernetes version.

Reply:
Extending the lifetime of a Kubernetes version will still be a very exceptional use-case that is under the control of operators. If they don't want to offer it, they won't do it.

The **new scenario** that you described will not be the default approach, it is still intended that Kubernetes versions expire globally. With `PrivateCloudProfile`s, operators just get the additional **option** to extended the expiration selectively.

Reply:
In an ad-hoc internal discussion with @timuthy @benedictweis @hendrikKahl and @dguendisch, we decide that it's better to have this feature.